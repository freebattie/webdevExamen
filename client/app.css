*{
    margin: 0;
    padding: 0;
    border: 0;
    box-sizing: border-box;

    font-family: 'Fira Sans', sans-serif;
}

.menu{
    margin-top: 40px;
    padding: 100px;
    display: flex;

    border: 1px solid black;
    flex-wrap: wrap;
    justify-content: center;

}
.dish{
    margin: 10px;
    width: 300px;
    border: solid 2px black;
    height: 200px;
}
footer{
    width: 80px;
}
.editItem{
    display: flex;
    width: 300px;
    flex-wrap: wrap;
}
.button{
    color: black;
    padding: 5px 2px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    cursor:pointer;
    margin: 10px;
    font-size: 16px;
    background-color: white;

    border: 2px solid black;
    width: 70px;
    border-radius: 5px;
    transition-duration: 0.4s;
}
.menuheader{
    margin-bottom: 20px;
}
.orders{
    display: flex;
    flex-wrap: wrap;
}
.order{

    height: 300px;
    border: 2px black solid;
    width: 300px;
    margin: 10px;
}
.items{
    width: 100px;
    margin: 10px;
}
.ordersheader{
    width: 100%;
}
.editform{
    width: 400px;
    border: 2px solid black;
}
.itemsToEdit{
    width: 100%;
    display: flex;
    flex-wrap: wrap;

}