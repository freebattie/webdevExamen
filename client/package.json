{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "scripts": {
    "test": "jest --coverage",
    "format": "prettier --write *.{jsx,js,tsx,ts,html}",
    "dev": "parcel index.html",
    "build": "npm install --include=dev && npm run build:parcel",
    "build:parcel": "parcel build index.html"
  },

  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@babel/preset-env": "^7.19.1",
    "@babel/preset-react": "7.18.6",
    "@types/react": "^18.0.21",
    "@types/react-dom": "^18.0.6",
    "babel-jest": "^29.0.3",
    "events": "^3.3.0",
    "jest": "^29.0.3",
    "jest-environment-jsdom": "^29.0.3",
    "parcel": "^2.7.0",
    "process": "^0.11.10",
    "rimraf": "^3.0.2"
  },
  "dependencies": {
    "@parcel/transformer-react-refresh-wrap": "^2.7.0",
    "@types/jest": "^29.0.3",
    "enzyme": "^3.11.0",
    "history": "^5.3.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.4.0"
  },
  "jest": {
    "testEnvironment": "jsdom",
    "collectCoverageFrom": [
      "**/*.jsx",
      "**/*.js",
      "!dist/**",
      "!coverage/**",
      "!app.jsx",
      "!index.jsx",
      "!lib/**"
    ]
  },
  "babel": {
    "presets": [
      "@babel/preset-env",
      [
        "@babel/preset-react",
        {
          "runtime": "automatic"
        }
      ]
    ]
  }
}
