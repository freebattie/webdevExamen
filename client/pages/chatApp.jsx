import React, { useState } from "react";
import { useNavigate } from "react-router-dom";

export function ChatApplication({ messages, onNewMessage }) {
  const [message, setMessage] = useState("");
  const navigate = useNavigate();
  function handleSubmit(e) {
    e.preventDefault();
    onNewMessage(message);
    setMessage("");
  }
  return (
    <>
      <header className={"footermove headerText"}>Modern Snack Chat</header>
      <main className={" border"}>
        {messages.map(({ message, user }, index) => (
          <div key={index}>
            <strong>{user}:</strong> {message}
          </div>
        ))}
      </main>
      <footer className={"footermove"}>
        <form className={"formsize"} onSubmit={handleSubmit}>
          <input
            className={"input2"}
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button className={"button"}>Send</button>
          <button className={"button"} onClick={(e) => navigate("/")}>
            back
          </button>
        </form>
      </footer>
    </>
  );
}
