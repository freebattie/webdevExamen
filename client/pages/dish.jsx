import { useEffect, useState } from "react";

export function Dish(props) {
  const [number, setNumber] = useState();
  let nextId = number;
  useEffect(() => {
    setNumber(props.val);
    const old = [...props.orders];
    old[0] = null;
    props.setOrders(old);
  }, []);
  async function handelAddToList() {
    setNumber((perv) => perv + 1);
    nextId++;
    setNumber(nextId);
    const item = {
      id: props.d.id,
      name: props.d.name,
      description: props.d.description,

      type: props.d.type,
      price: props.d.price,
      size: nextId,
    };
    const old = [...props.orders];
    old[props.index] = item;
    props.setOrders(old);
  }
  async function handelRemoveToList() {
    if (number > 0) {
      nextId--;
      setNumber(nextId);
      const item = {
        id: props.d.id,
        name: props.d.name,
        price: props.d.price,
        size: nextId,
      };
      const old = [...props.orders];
      old[props.d.id] = item;
      props.setOrders(old);
    }
  }

  return (
    <div className={"dish"}>
      <div>name :{props.d.name}</div>
      <div>price:{props.d.price}Nok</div>
      <div>description:{props.d.description}</div>

      <div>type Dish :{props.d.type}</div>
      <div>order :{number}</div>
      <button className={"button"} onClick={() => handelAddToList()}>
        add
      </button>
      <footer>
        <button className={"button"} onClick={() => handelRemoveToList()}>
          remove
        </button>
      </footer>
    </div>
  );
}
