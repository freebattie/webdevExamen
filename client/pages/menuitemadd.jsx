import { useContext, useEffect, useState } from "react";
import { Appcontext } from "../lib/appcontext.jsx";
import { useNavigate } from "react-router-dom";
import { LogOut } from "../Components/logout";

export function MenuItemAdd(props) {
  const { addDish } = useContext(Appcontext);
  const { removeDish } = useContext(Appcontext);
  const [name, setName] = useState("");
  const [id, setId] = useState("");
  const [price, setPrice] = useState(0);
  const [description, setDescription] = useState("");
  const [type, setType] = useState("");
  const navigate = useNavigate();

  async function handelSubmit(e) {
    e.preventDefault();
    const dish = {
      _id: id,
      id: id,
      name,
      price,

      description,
      type,
    };
    try {
      await addDish(dish);

      navigate("/");
    } catch (e) {
      props.setError(e);
      navigate("/error");
    }
  }

  async function handelBackClick(e) {
    navigate("/editmenu");
  }
  async function handelHomeClick(e) {
    navigate("..");
  }
  return (
    <div className={"editform"}>
      <form onSubmit={handelSubmit}>
        <div className={"editItem"}>
          <label>
            name:
            <input
              type="text"
              value={name}
              onChange={(e) => setName(e.target.value)}
            />
          </label>
          <label>
            price:
            <input
              type="number"
              value={price}
              onChange={(e) => setPrice(e.target.value)}
            />
          </label>
          <label>
            description:
            <input
              type="text"
              value={description}
              onChange={(e) => setDescription(e.target.value)}
            />
          </label>
          <label>
            type:
            <input
              type="text"
              value={type}
              onChange={(e) => setType(e.target.value)}
            />
          </label>
        </div>
        <button className={"button"}>save</button>
      </form>
      <button className={"button"} onClick={(e) => handelBackClick(e)}>
        back
      </button>
      <button className={"button"} onClick={(e) => handelHomeClick(e)}>
        home
      </button>
      <LogOut />
    </div>
  );
}
