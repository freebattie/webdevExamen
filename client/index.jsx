import { createRoot } from "react-dom/client";

import { App } from "./app";

// get the root element in index.html
const container = document.getElementById("app");
//creat a react root from that id
const root = createRoot(container);

root.render(<App />);
