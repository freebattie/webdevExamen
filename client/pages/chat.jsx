import { useEffect, useState } from "react";
import { ChatApplication } from "./chatApp.jsx";
const initialMessages = [];

function UserRegistrationForm({ onUsername }) {
  const [username, setUsername] = useState("");
  function handleSubmit(event) {
    event.preventDefault();
    onUsername(username);
  }
  return (
    <center>
      <form className={"loginChat"} onSubmit={handleSubmit}>
        <div className={"loginChatfld"}>
          <label>
            Username:
            <input
              className={"input"}
              value={username}
              onChange={(e) => setUsername(e.target.value)}
            />
          </label>
          <button className={"button"}>Submit</button>
        </div>
      </form>
    </center>
  );
}
export function Chat() {
  const [user, setUser] = useState();
  const [messages, setMessages] = useState(initialMessages);
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://" + window.location.host + "/api/chat");
    ws.onopen = (event) => {
      console.log("Opened", event);
    };
    ws.onmessage = (event) => {
      console.log("message", event);
      const { user, message } = JSON.parse(event.data);
      setMessages((messages) => [...messages, { message, user }]);
    };
    setWs(ws);
  }, []);

  function handleNewMessage(message) {
    ws.send(JSON.stringify({ message, user }));
  }

  if (!user) {
    return <UserRegistrationForm onUsername={setUser} />;
  }

  return (
    <ChatApplication messages={messages} onNewMessage={handleNewMessage} />
  );
}
