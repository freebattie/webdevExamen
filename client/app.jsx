import { BrowserRouter, Routes, Route } from "react-router-dom";
import { FrontPage } from "./pages/frontpage.jsx";
import { EditUsers } from "./pages/edituser.jsx";
import { ErrorUser } from "./pages/erroruser.jsx";
import { Login } from "./pages/login.jsx";
import { NewUser } from "./pages/newuser.jsx";

import { useEffect, useState } from "react";
import { MyOrders } from "./pages/myorders.jsx";
import { Error } from "./pages/error.jsx";
import { EditMenu } from "./pages/editmenu.jsx";

import { MenuItemAdd } from "./pages/menuitemadd";
import { ChatApplication } from "./pages/chatApp.jsx";
import { Chat } from "./pages/chat";

function AddMenu({ setError }) {
  return (
    <div>
      <MenuItemAdd setError={setError} />
    </div>
  );
}

export function App() {
  const [error, setError] = useState();
  return (
    <BrowserRouter>
      <Routes>
        <Route path={"/"} element={<FrontPage setError={setError} />} />
        <Route path={"/login"} element={<Login setError={setError} />} />
        <Route path={"/register"} element={<NewUser setError={setError} />} />
        <Route path={"/edit"} element={<EditUsers />} />
        <Route path={"/error"} element={<Error error={error} />} />
        <Route path={"/errorUser"} element={<ErrorUser />} />
        <Route path={"/Myorders"} element={<MyOrders />} />
        <Route path={"/editmenu"} element={<EditMenu setError={setError} />} />
        <Route path={"/addmenu"} element={<AddMenu setError={setError} />} />
        <Route path={"/chat"} element={<Chat />} />
      </Routes>
    </BrowserRouter>
  );
}
