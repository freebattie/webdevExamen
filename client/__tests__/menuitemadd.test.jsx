import * as React from "react";

import { createRoot } from "react-dom/client";
import { act, Simulate } from "react-dom/test-utils";

import { MenuItemAdd } from "../pages/menuitemadd";
import { MemoryRouter } from "react-router-dom";
global.IS_REACT_ACT_ENVIRONMENT = true;

describe("app", () => {
  it("should rener App", async function () {
    const element = document.createElement("div");

    //creat a React root from that id
    const root = createRoot(element);
    const setError = jest.fn();
    await act(async () => {
      root.render(
        <MemoryRouter>
          <MenuItemAdd setError={setError} />
        </MemoryRouter>
      );
    });
    await act(async () => {
      global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    });
    await act(async () => {
      await Simulate.change(
        await element.querySelector("form div label:nth-of-type(1) input"),
        {
          target: { value: "Petter petter" },
        }
      );
      await Simulate.change(
        await element.querySelector("form div label:nth-of-type(2) input"),
        {
          target: { value: "Petter petter" },
        }
      );
      await Simulate.change(
        await element.querySelector("form div label:nth-of-type(3) input"),
        {
          target: { value: "Petter petter" },
        }
      );
      await Simulate.change(
        await element.querySelector("form div label:nth-of-type(4) input"),
        {
          target: { value: "Petter petter" },
        }
      );
    });
    await act(async () => {
      await Simulate.submit(await element.querySelector("form"));
    });
    expect(element).toMatchSnapshot();
    await act(async () => {});
  });
});
