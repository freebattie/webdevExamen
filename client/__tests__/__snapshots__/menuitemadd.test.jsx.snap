// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app should rener App 1`] = `
<div>
  <div
    class="editform"
  >
    <form>
      <div
        class="editItem"
      >
        <label>
          name:
          <input
            type="text"
            value="Petter petter"
          />
        </label>
        <label>
          price:
          <input
            type="number"
            value="Petter petter"
          />
        </label>
        <label>
          description:
          <input
            type="text"
            value="Petter petter"
          />
        </label>
        <label>
          type:
          <input
            type="text"
            value="Petter petter"
          />
        </label>
      </div>
      <button
        class="button"
      >
        save
      </button>
    </form>
    <button
      class="button"
    >
      back
    </button>
    <button
      class="button"
    >
      home
    </button>
    <button
      class="button"
    >
      Logout
    </button>
  </div>
</div>
`;
